pkg/flow/internal/controller/loader.go

```diff
@@ -564,84 +568,106 @@ func (l *Loader) OriginalGraph() *dag.Graph {
-			// EvaluateDependencies had its exports changed and none of its input
-			// arguments will need re-evaluation.
-			return nil
+		// Submit the node for asynchronous evaluation.
+		err := l.workerPool.SubmitWithKey(n.NodeID(), l.concurrentEvalFn(n, dependantCtx, tracer, parent))
+		if err != nil {
+			// The error typically means that the workerPool queue is full. This could mean we have too many components
+			// and the agent cannot keep up with the evaluation. To degrade gracefully, we log the error and continue.
+			// Some evaluations may be missed.
+			level.Error(l.log).Log("msg", "failed to submit node for evaluation - the agent is likely "+
+				"overloaded and cannot keep up with evaluating components - some updates may be dropped",
+				"node_id", n.NodeID(), "err", err)
+			span.SetStatus(codes.Error, err.Error())
 		}
+		span.End()
+	}
+
+	// Report queue size metric.
+	l.cm.evaluationQueueSize.Set(float64(l.workerPool.QueueSize()))
+}
 
+// concurrentEvalFn returns a function that evaluates a node and updates the cache. This function can be submitted to
+// a worker pool for asynchronous evaluation.
+func (l *Loader) concurrentEvalFn(n dag.Node, spanCtx context.Context, tracer trace.Tracer, parent *ComponentNode) func() {
+	return func() {
+		start := time.Now()
+		l.cm.dependenciesWaitTime.Observe(time.Since(parent.lastUpdateTime.Load()).Seconds())
 		_, span := tracer.Start(spanCtx, "EvaluateNode", trace.WithSpanKind(trace.SpanKindInternal))
 		span.SetAttributes(attribute.String("node_id", n.NodeID()))
 		defer span.End()
 
-		start := time.Now()
 		defer func() {
-			level.Info(logger).Log("msg", "finished node evaluation", "node_id", n.NodeID(), "duration", time.Since(start))
+			duration := time.Since(start)
+			level.Info(l.log).Log("msg", "finished node evaluation", "node_id", n.NodeID(), "duration", duration)
+			l.cm.componentEvaluationTime.Observe(duration.Seconds())
 		}()
 
-		// Record how long components wait to be evaluated after their dependency is updated
-		if parentCmp, ok := parent.(*ComponentNode); ok {
-			l.cm.dependenciesWaitTime.Observe(time.Since(parentCmp.lastUpdateTime.Load()).Seconds())
-		}
-
 		var err error
 		switch n := n.(type) {
 		case BlockNode:
-			err = l.evaluate(logger, n)
+			ectx := l.cache.BuildContext()
+			evalErr := n.Evaluate(ectx)
+
+			// Only obtain loader lock after we have evaluated the node, allowing for concurrent evaluation.
+			l.mut.RLock()
+			defer l.mut.RUnlock()
+			err = l.postEvaluate(l.log, n, evalErr)
+
+			// Additional post-evaluation steps necessary for module exports.
 			if exp, ok := n.(*ExportConfigNode); ok {
 				l.cache.CacheModuleExportValue(exp.Label(), exp.Value())
 			}
+			if l.globals.OnExportsChange != nil && l.cache.ExportChangeIndex() != l.moduleExportIndex {
+				l.globals.OnExportsChange(l.cache.CreateModuleExports())
+				l.moduleExportIndex = l.cache.ExportChangeIndex()
```

Comment:
Is it safe to modify this value with only a read lock?

Reply:
Great question! I do _think_ it was safe, but certainly not obvious and it could generate unnecessary calls to `OnExportsChange` when a stale `moduleExportIndex` was read. So, to improve clarity, I've modified this code to upgrade to a write lock when this update is needed. I will be doing some of the tests again to verify.